src/bun.js/webcore/ObjectURLRegistry.zig

```diff
@@ -0,0 +1,138 @@
+const std = @import("std");
+const bun = @import("root").bun;
+const JSC = bun.JSC;
+const UUID = bun.UUID;
+const assert = bun.assert;
+const ObjectURLRegistry = @This();
+
+lock: bun.Lock = bun.Lock.init(),
+map: std.AutoHashMap(UUID, *RegistryEntry) = std.AutoHashMap(UUID, *RegistryEntry).init(bun.default_allocator),
+
+pub const RegistryEntry = struct {
+    blob: JSC.WebCore.Blob,
+
+    pub usingnamespace bun.New(@This());
+    pub fn init(blob: *const JSC.WebCore.Blob) *RegistryEntry {
+        return RegistryEntry.new(.{
+            .blob = blob.dupeWithContentType(true),
+        });
+    }
+
+    pub fn deinit(this: *RegistryEntry) void {
+        this.blob.deinit();
+        this.destroy();
+    }
+};
+
+pub fn register(this: *ObjectURLRegistry, vm: *JSC.VirtualMachine, blob: *const JSC.WebCore.Blob) UUID {
+    const uuid = vm.rareData().nextUUID();
+    const entry = RegistryEntry.init(blob);
+
+    this.lock.lock();
+    defer this.lock.unlock();
+    this.map.put(uuid, entry) catch bun.outOfMemory();
+    return uuid;
+}
+
+pub fn singleton() *ObjectURLRegistry {
+    const Singleton = struct {
+        pub var registry: ObjectURLRegistry = undefined;
+        pub var once = std.once(get);
+
+        fn get() void {
+            registry = .{};
+        }
+    };
+
+    Singleton.once.call();
+
+    return &Singleton.registry;
+}
+
+fn getDupedBlob(this: *ObjectURLRegistry, uuid: *const UUID) ?JSC.WebCore.Blob {
+    this.lock.lock();
+    defer this.lock.unlock();
+    const entry = this.map.get(uuid.*) orelse return null;
+    return entry.blob.dupeWithContentType(true);
+}
+
+fn uuidFromPathname(pathname: []const u8) ?UUID {
+    return UUID.parse(pathname) catch return null;
+}
+
+pub fn resolveAndDupe(this: *ObjectURLRegistry, pathname: []const u8) ?JSC.WebCore.Blob {
+    const uuid = uuidFromPathname(pathname) orelse return null;
+    this.lock.lock();
+    defer this.lock.unlock();
+    const entry = this.map.get(uuid) orelse return null;
+    return entry.blob.dupeWithContentType(true);
+}
+
+pub fn revoke(this: *ObjectURLRegistry, pathname: []const u8) void {
+    const uuid = uuidFromPathname(pathname) orelse return;
+    this.lock.lock();
+    defer this.lock.unlock();
+    const entry = this.map.fetchRemove(uuid) orelse return;
+    entry.value.deinit();
+}
+
+pub fn has(this: *ObjectURLRegistry, pathname: []const u8) bool {
+    const uuid = uuidFromPathname(pathname) orelse return false;
+    this.lock.lock();
+    defer this.lock.unlock();
+    return this.map.contains(uuid);
+}
+
+export fn Bun__createObjectURL(globalObject: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) callconv(.C) JSC.JSValue {
+    const arguments = callframe.arguments(1);
+    if (arguments.len < 1) {
+        globalObject.throwNotEnoughArguments("createObjectURL", 1, arguments.len);
```

Comment:
we should add some tests covering all exceptions in `ObjectURLRegistry` and see if nodejs equivalent throws the same error in this situations.